### ISSUE TITLE
Make "required" validation message configurable

### ISSUE THREAD

**Issue by svenstaro** [marshmallow-code/marshmallow#78]
I find the current required=bool lacking because it doesn't allow me to customize its error message. It would be better if required were implemented as a validator itself. This is especially useful for providing semantic per-field required help or for translating the message.

I've been looking at #76 too and this could be implemented in the same go. Also, it would stream line some code as currently you are handling required separately from everything else even though in reality it could be just another validator. Thoughts?

**Comment by sloria** [comment]
The validation for `required` is handled separately because it must precede execution of the `validators` callables. If a required value is not passed, validation should fail fast (as there is no value to pass to the rest of the validators). 

I agree that it would be useful to allow a per-field message for `required` validation. Perhaps we could allow `required` to be a `bool` _or_ a string which is the validation message.

``` python
class MySchema(Schema):
    num = fields.Int(required='Datos que faltan para el campo requerido')

MySchema().validate({})
# {'num': 'Datos que faltan para el campo requerido'}
```

**Comment by svenstaro** [comment]
I see, that makes sense. I like your solution, we can use that and it's certainly simple enough.

**Comment by sloria** [comment]
Thanks for your feedback. Pull requests welcome!

**Comment by 0xDCA** [comment]
I find it clearer if there is a different required_error parameter instead of setting required to a string. What do you think?

I think I can do the PR anyway.

**Comment by sloria** [comment]
@0xDCA Both alternatives come with different tradeoffs, each adding complexity in different ways. I slightly prefer to just have just the `required` parameter (that can be a `bool` or a `str`) because it reduces the number of `Field` parameters and avoids having "dependent" parameters (e.g. what happens if a user passes `required_error` but sets `required=False`?).

If you would like to send a PR adding this functionality, I would gladly review and merge it.

**Comment by 0xDCA** [comment]
Setting required_error without setting required to True would just do that: setting the error message. It's exactly the same effect when you set a custom error message (for any other validator) and the validation never fails (for whatever reason). 

As the current code may expect required to be a bool, something like "if self.required" would fail for empty strings (though setting an empty error message does not have any sense).

I think I can send the PR later today (with the approach that you decide).

**Comment by sloria** [comment]
I think I still prefer to have the one `required` parameter. Let's go with that for now.

> As the current code may expect required to be a bool, something like "if self.required" would fail for empty strings (though setting an empty error message does not have any sense).

This can be changed to check `if self.required is True` if need be.

**Comment by 0xDCA** [comment]
PR sent: https://github.com/marshmallow-code/marshmallow/pull/122

**Comment by sloria** [comment]
Closed by #122
